### PR TITLE
Shortcuts for Android 7.1

### PR DESCRIPTION
**Pull request by Dafnik**
### First time contributor checklist

- [x] I have read [how to contribute](https://github.com/WhisperSystems/Signal-Android/blob/master/CONTRIBUTING.md) to this project
- [x] I have signed the [Contributor License Agreement](https://whispersystems.org/cla/)

### Contributor checklist

- [x] I am following the [Code Style Guidelines](https://github.com/WhisperSystems/Signal-Android/wiki/Code-Style-Guidelines)
- [x] I have tested my contribution on these devices:
 * Oneplus 3, 7.1.1
- [x] My contribution is fully baked and ready to be merged as is
- [x] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)
- [x] I have made the choice whether I want the [BitHub reward](https://github.com/WhisperSystems/Signal-Android/wiki/BitHub-Rewards) or not by omitting or adding the word `FREEBIE` in the commit message of my first commit

----------

### Description

Android 7.1 adds a new launcher feature, the shortcuts. They are implemented for making life easier. So you have two options. Firstly, you can create a new conversation. And secondly, you can get into a recent conversation with your contacts or groups.
Here is a scrennshot, so you know how it looks like:
shortcuts

### CODE REVIEW

**Comment by chrihu** [comment]
Please take a look at Googles [guidelines](https://developer.android.com/guide/topics/ui/shortcuts.html#best-practices) for shortcuts. 

Following the guidelines I would drop the "New **g**roup" shortcut entirely and regarding the icons I have attached a proper version for "New **c**onversation". (**lower case**)
[ic_shortcut_create.xml.zip](https://github.com/WhisperSystems/Signal-Android/files/731277/ic_shortcut_create.xml.zip)

Screenshot:
screenshot2

**Comment by Dafnik** [comment]
Why would you drop "New group" shortcut?

The shortcuts design looks different to yours because I use a different launcher.

**Comment by p4nci** [comment]
@Dafnik [Googles Best Practices for App Shortcuts](https://developer.android.com/guide/topics/ui/shortcuts.html#best-practices) says:
```
Publish only four distinct shortcuts
```

**Comment by Dafnik** [comment]
Okay

**Comment by chrihu** [comment]
I am talking about the icon not the shortcuts.
Google is very specific about the design and file format the shortcut icon should have as you can read [here](https://material.io/guidelines/style/icons.html#icons-app-shortcut-icons).
I have attached one in my last post that you can use if you like.

**Comment by Dafnik** [comment]
@chrihu, I've added the new icon.

**Comment by Dafnik** [comment]
What do you think @moxie0?

File: res/values/strings.xml [hunk 1]
```diff
@@ -1292,6 +1292,8 @@
     <!-- transport_selection_list_item -->
     <string name="transport_selection_list_item__transport_icon">Transport icon</string>
 
+    <!-- Shortcuts -->
+    <string name="shortcut_new_conversation">New conversation</string>
```
Comment (moxie0) [inline res/values/strings.xml]: looks like this should be "shortcuts__new_conversation" by convention

File: src/org/thoughtcrime/securesms/service/MessageRetrievalService.java [hunk 2]
```diff
@@ -124,12 +128,14 @@ public IBinder onBind(Intent intent) {
...
 
   private synchronized void incrementActive() {
+    createShortcuts();
     activeActivities++;
     Log.w(TAG, "Active Count: " + activeActivities);
     notifyAll();
   }
 
   private synchronized void decrementActive() {
+    createShortcuts();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/service/MessageRetrievalService.java]: i don't think this belongs here, why would the message retrieval service know anything about shortcuts?

File: src/org/thoughtcrime/securesms/service/MessageRetrievalService.java [hunk 3]
```diff
@@ -188,4 +194,10 @@ public static void registerActivityStopped(Context activity) {
   public static @Nullable SignalServiceMessagePipe getPipe() {
     return pipe;
   }
+
+  private void createShortcuts() {
+    if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
+      new DynamicShortcutCreator(MessageRetrievalService.this).buildShortcuts();
+    }
+  }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/service/MessageRetrievalService.java]: same, this feels very out of place, classes should only know about things that are their responsibilities

File: src/org/thoughtcrime/securesms/database/ThreadDatabase.java [hunk 4]
```diff
@@ -354,6 +354,11 @@ public Cursor getDirectShareList() {
     return db.query(TABLE_NAME, null, null, null, null, null, DATE + " DESC");
   }
 
+  public Cursor getShortcutList() {
+    SQLiteDatabase db = databaseHelper.getReadableDatabase();
+    return db.query(TABLE_NAME, null, null, null, null, null, DATE + " DESC");
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/database/ThreadDatabase.java]: do you really need to build results for every single thread in the db in order to create shortcuts? seems like you should have a limit in here somewhere

File: src/org/thoughtcrime/securesms/service/MessageRetrievalService.java [hunk 5]
```diff
@@ -188,4 +194,10 @@ public static void registerActivityStopped(Context activity) {
   public static @Nullable SignalServiceMessagePipe getPipe() {
     return pipe;
   }
+
+  private void createShortcuts() {
+    if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
+      new DynamicShortcutCreator(MessageRetrievalService.this).buildShortcuts();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/service/MessageRetrievalService.java]: you're doing a ton of i/o on the UI thread

File: src/org/thoughtcrime/securesms/DynamicShortcutCreator.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
+package org.thoughtcrime.securesms;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/DynamicShortcutCreator.java]: I haven't reviewed the rest of this file, but the entire thing looks to be 4-space indents. please read the style notes for this repository

**Comment by moxie0** [comment]
please reopen when updated, thanks!

**Comment by Stanzilla** [comment]
@Dafnik are you still working on this?

**Comment by Dafnik** [comment]
No. I don't have time to take a closer look at the project do implement it the "right" place.

**Pull request closed without merging**